coreax/metrics.py

```diff
@@ -428,3 +429,200 @@ def sum_weighted_pairwise_distances(
+            \{(x_i, y_i)\}_{i=1}^n` of ``n`` pairs with :math:`x\in\mathbb{R}^d` and
+            :math:`y\in\mathbb{R}^p`, responses should be concatenated after the
+            features
+        :param dataset_2: Dataset :math:`\mathcal{D}^{(2)} = \{(\tilde{x}_i,
+            \tilde{y}_i)\}_{i=1}^n` of ``m`` pairs with :math:`\tilde{x}\in\mathbb{R}^d`
+            and :math:`\tilde{y}\in\mathbb{R}^p`, responses should be
+            concatenated after the features
+        """
+        # Extract and format features and responses from D1 and D2
+        x1 = jnp.atleast_2d(dataset_1[:, : self.num_feature_dimensions])
+        y1 = jnp.atleast_2d(dataset_1[:, self.num_feature_dimensions :])
+        x2 = jnp.atleast_2d(dataset_2[:, : self.num_feature_dimensions])
+        y2 = jnp.atleast_2d(dataset_2[:, self.num_feature_dimensions :])
+
+        # Compute feature kernel gramians
+        feature_gramian_1 = self.feature_kernel.compute(x1, x1)
+        feature_gramian_2 = self.feature_kernel.compute(x2, x2)
+        cross_feature_gramian = self.feature_kernel.compute(x2, x1)
+
+        # Compute response kernel gramians
+        response_gramian_1 = self.response_kernel.compute(y1, y1)
+        response_gramian_2 = self.response_kernel.compute(y2, y2)
+        cross_response_gramian = self.response_kernel.compute(y1, y2)
+
+        # Invert feature kernel gramians
+        identity_1 = jnp.eye(feature_gramian_1.shape[0])
+        inverse_feature_gramian_1 = coreax.util.invert_regularised_array(
+            array=feature_gramian_1,
+            regularisation_parameter=self.regularisation_parameters[0].item(),
+            identity=identity_1,
+        )
+
+        identity_2 = jnp.eye(feature_gramian_2.shape[0])
+        inverse_feature_gramian_2 = coreax.util.invert_regularised_array(
+            array=feature_gramian_2,
+            regularisation_parameter=self.regularisation_parameters[1].item(),
+            identity=identity_2,
+        )
+
+        # Compute each term in the CMMD
+        term_1 = (
+            inverse_feature_gramian_1
+            @ response_gramian_1
+            @ inverse_feature_gramian_1
+            @ feature_gramian_1
+        )
+        term_2 = (
+            inverse_feature_gramian_2
+            @ response_gramian_2
+            @ inverse_feature_gramian_2
+            @ feature_gramian_2
+        )
+        term_3 = (
+            inverse_feature_gramian_1
+            @ cross_response_gramian
+            @ inverse_feature_gramian_2
+            @ cross_feature_gramian
+        )
+
+        # Compute CMMD
+        squared_result = jnp.trace(term_1) + jnp.trace(term_2) - 2 * jnp.trace(term_3)
+        if squared_result < 0:
+            warn(
+                f"Squared CMMD ({round(squared_result.item(), 4)}) is negative,"
+                + " increase precision threshold or regularisation strength.",
+                Warning,
+                stacklevel=1,
```

Comment:
Stacklevel probably wants to be 2. We have another example somewhere in the code where this same change should be made.

Reply:
Will change!